layers/eight_mile/pytorch/layers.py

```diff
@@ -2669,6 +2669,61 @@ def _attention(
         return F.softmax(scores, dim=-1)
 
 
+class SeqScaledMaskedRelativeAttention(SequenceSequenceRelativeAttention):
+    """This class implements masked relative attention, i.e. preventing attention beyond rpr_k. For efficiency,
+    _attention and _update are implemented in a different way."""
+    def __init__(self, pdrop: float = 0.1, **kwargs):
+        super().__init__(pdrop=pdrop, **kwargs)
+
+    def _unfold_mask(self, mask, batchsz, rpr_k):
+        """Transform mask into the unfolded format."""
+        window_len = 2 * rpr_k + 1
+        T = mask.shape[3]
+        mask = mask.expand(batchsz, 1, T, T)  # expand sequence/subsequent mask into a uniform dim
+        mask = F.pad(mask, [rpr_k, rpr_k])  # pad both sides with rpr_k, [B, 1, T, T + 2*rpr_k]
+        seq = torch.arange(T + 2 * rpr_k)
+        indices = seq.unfold(0, window_len, 1)  # indices of a sliding window, [T, W]
+        indices = indices.unsqueeze(0).unsqueeze(0).expand(batchsz, 1, T, window_len).to(mask.device)
+        return torch.gather(mask, -1, indices)  # [B, 1, T, W]):
+
+    def _attention(
+            self, query: torch.Tensor, key: torch.Tensor, rpr_key: torch.Tensor, mask: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        """Implementation of attention considering RA masking: using torch.Tensor.unfold to create an extra dimension
+        representing the sliding window. Then when applying matmul, Q, K, V share the same T dimension.
+
+        :param query: a query for alignment. Can come from self in case of self-attn or decoder in case of E/D
+        :param key: a set of keys from encoder or self
+        :param mask: masking (for destination) to prevent seeing what we shouldnt
+        :param rpr_key: tensor of the rpr_key embeddings [window_sz, d_k]
```

Comment:
comments refer to both `window_sz` and `W` but it looks like they are the same.  Lets pick on and update the comments to be consistent